librz/io/p/io_srec.c

```diff
@@ -0,0 +1,798 @@
+			str = eol;
+			break;
+
+		case 6: // same with S5 but on 24 bits instead of 16
+
+			// S / 6 / (addres+data+checksum) bytes / counter / - / checksum
+			l = sscanf(str + 4, "%06x", &addr_tmp);
+			addr_tmp &= 0xffffff;
+
+			eol = strchr(str + 1, 'S');
+			if (eol) {
+				*eol = 0;
+			}
+
+			cksum = bc;
+			cksum += addr_tmp >> 8;
+			cksum += addr_tmp >> 16;
+			cksum += addr_tmp & 0xff;
+
+			if ((next_addr != addr_tmp) || ((sec_size + bc) > SEC_MAX)) {
+				// previous block is not contiguous, or
+				// section buffer is full => write a sparse chunk
+				if (sec_size && sec_size < UT16_MAX) {
+					if (rz_buf_write_at(rbuf, sec_start, sec_tmp, (st32)sec_size) != sec_size) {
+						eprintf("sparse buffer problem, giving up\n");
+						goto fail;
+					}
+				}
+				// advance cursor, reset section
+				sec_start = segreg + addr_tmp;
+				next_addr = addr_tmp;
+				sec_size = 0;
+			}
+			sec_size += bc;
+			next_addr += bc;
+			if (eol) {
+				// checksum
+				if (sscanf(str + 10 + (i * 2), "%02x", &byte) != 1) {
+					eprintf("unparsable data !\n");
+					goto fail;
+				}
+				cksum = ~cksum; // this might be useless cause cksum is ut8
+				if (cksum != byte) {
+					eprintf("Checksum failed (got %02x expected %02x)\n", byte, cksum);
+					goto fail;
+				}
+				*eol = 'S';
+			}
+			str = eol;
+			break;
+
+		case 8: // same with S9 but on 24 bits instead of 16
+
+			// S / 8 / (addres+data+checksum) bytes / adress / - / checksum
+			l = sscanf(str + 4, "%06x", &addr_tmp);
+			addr_tmp &= 0xffffff;
+
+			eol = strchr(str + 1, 'S');
+			if (eol) {
+				*eol = 0;
+			}
+
+			cksum = bc;
+			cksum += addr_tmp >> 8;
+			cksum += addr_tmp >> 16;
+			cksum += addr_tmp & 0xff;
+
+			if ((next_addr != addr_tmp) || ((sec_size + bc) > SEC_MAX)) {
+				// previous block is not contiguous, or
+				// section buffer is full => write a sparse chunk
+				if (sec_size && sec_size < UT16_MAX) {
+					if (rz_buf_write_at(rbuf, sec_start, sec_tmp, (st32)sec_size) != sec_size) {
+						eprintf("sparse buffer problem, giving up\n");
+						goto fail;
+					}
+				}
+				// advance cursor, reset section
+				sec_start = segreg + addr_tmp;
+				next_addr = addr_tmp;
+				sec_size = 0;
+			}
+			sec_size += bc;
+			next_addr += bc;
+			if (eol) {
+				// checksum
+				if (sscanf(str + 10 + (i * 2), "%02x", &byte) != 1) {
+					eprintf("unparsable data !\n");
+					goto fail;
+				}
+				cksum = ~cksum;
+				if (cksum != byte) {
+					eprintf("Checksum failed (got %02x expected %02x)\n", byte, cksum);
+					goto fail;
+				}
+				*eol = 'S';
+			}
+			str = eol;
+			break;
+
+		case 3: // 32 bit adress, with data field!, kinda same with S1
+
+			// S / 3 / (addres+data+checksum) bytes / address / data / checksum
+			l = sscanf(str + 4, "%08x", &addr_tmp);
+			addr_tmp &= 0xffffffff;
+
+			eol = strchr(str + 1, 'S');
+			if (eol) {
+				*eol = 0;
+			}
+
+			cksum = bc;
+			cksum += addr_tmp >> 8;
+			cksum += addr_tmp >> 16;
+			cksum += addr_tmp >> 32;
```

Comment:
this shift is wrong. are yo sure that isn't 24 ?

Reply:
Oh wait it is 24, my bad